File: 2024/day05/day05.py
from collections import defaultdict
from pprint import pprint
import sys

rules = []
queries = []
rules_done = False

filename = "problem.in"
filename = "example.in"
filename = sys.argv[1]

with open(filename) as f:
    for l in f.readlines():
        l = l.strip()
        if not l:
            rules_done = True
        elif not rules_done:
            rules.append([int(a) for a in l.split("|")])
        else:
            queries.append([int(a) for a in l.split(",")])



graph = dict()
inv_graph = defaultdict(list)
in_degrees = defaultdict(lambda: 0)
ssc = dict()
"""strongly connected components. If ssc[u] == root, then u is in component number `root`"""

def find_ssc():
    L = []
    visited = set()

    def visit(u):
        if u in visited:
            return
        visited.add(u)
        for v in graph[u]:
            visit(v)
        L.insert(0, u)

    for n in graph:
        visit(n)

    def assign(u, root):
        if u in ssc:
            return
        ssc[u] = root
        for v in inv_graph[u]:
            assign(v, root)

    pprint(L)
    for u in L:
        assign(u, u)
    pprint(ssc)



def one():
    for src, dst in rules:
        if src not in graph:
            graph[src] = []
        graph[src].append(dst)
        inv_graph[dst].append(src)
        in_degrees[dst] += 1
        if src not in in_degrees:
            in_degrees[src] = 0
        if dst not in graph:
            graph[dst] = []

    pprint(graph)
    pprint(inv_graph)
    pprint(in_degrees)
    find_ssc()

def one_brute():
    for src, dst in rules:
        if src not in graph:
            graph[src] = set()
        if dst not in graph:
            graph[dst] = set()
        graph[src].add(dst)

    pprint(graph)

one_brute()

total = 0
for q in queries:
    print(q)
    ok = True
    for i in range(len(q) - 1):
        print(q[i], q[i+1])
        if q[i] in graph[q[i+1]]:
           print(q[i+1], "<", q[i])
           ok = False
           break
    if ok:
        total += q[len(q) // 2]
print(total)

File: 2024/day05/day05.go
package main

import (
    "fmt"
    "log"
    "os"

    "helpers/aoc"
)

var permament [99]bool
var temporary [99]bool
var unmarked map[int]bool = make(map[int]bool)

func visit(n int, graph map[int][]int) {
    if permament[n] {
        return
    }
    fmt.Printf("visit %v\n", n)
    if temporary[n] {
        log.Fatal("cycle")
    }
    delete(unmarked, n)
    temporary[n] = true
    for _, m := range graph[n] {
        visit(m, graph)
    }
    permament[n] = true
    fmt.Println(n)
}

func one(rules, queries [][]int) int {
    // Pomysł jest taki, że sortujemy topologicznie wierzchołki
    // I potem porównujemy query do posortowanych wierzchołków i sprawdzamy czy to podciąg
    // To będzie działać jeśli WSZYSTKIE wierzchołki będą połączone
    // Czyli będzie ile krawędzi? N*(N-1)/2 ?
    // Przykład na to wskazuje, ale opis nie do końca
    graph := make(map[int][]int)
    inDegrees := make(map[int]int)
    for _, rule := range rules {
        src, dst := rule[0], rule[1]
        graph[src] = append(graph[src], dst)
        inDegrees[dst] += 1
        if _, exists := inDegrees[src]; !exists {
            inDegrees[src] = 0
        }
    }
    for n, _ := range graph {
        unmarked[n] = true
    }
    fmt.Println(len(rules))
    fmt.Println(len(graph))
    //fmt.Println(graph)
    fmt.Println(inDegrees)

    for {
        if len(unmarked) <= 0 {
            return 0
        }
        for n, _ := range unmarked {
            visit(n, graph)
            break
        }
    }
    return 0
}


func two() int {
    return 0
}


func main() {
    log.SetPrefix("")
    log.SetFlags(0)
    if len(os.Args) != 2 {
        log.Fatal("Usage: ", os.Args[0], " <filename>\n")
    }
    inName := os.Args[1]
    fmt.Printf("Reading from: %v\n", inName)

    lines := aoc.ReadLines(inName)
    rulesRead := false
    var rules [][]int = make([][]int, 0)
    var rule []int
    var queries [][]int = make([][]int, 0)
    var query []int
    for _, l := range lines {
        if l == "" {
            rulesRead = true
        } else if !rulesRead {
            rule = aoc.AllInts(l, "|")
            rules = append(rules, rule)
        } else {
            query = aoc.AllInts(l, ",")
            queries = append(queries, query)
            //fmt.Println(query)
        }
    }

    ansOne := one(rules, queries)
    ansTwo := two()
    fmt.Printf("Part 1: %v\n", ansOne)
    fmt.Printf("Part 2: %v\n", ansTwo)
}
